GitHub runner webhook router create and add charm architecture page and diagram (#72)
canonical/github-runner-webhook-router: charm/docs/charm-architecture.md

# Charm architecture

> The GitHub Runner Webhook Router is a Flask application tha provides a webhook router for self-hosted GitHub Actions runners. The [GitHub Runner Webhook Router charm](https://github.com/canonical/github-runner-webhook-router/) was developed using the [12-factor Flask framework](https://canonical-charmcraft.readthedocs-hosted.com/en/stable/reference/extensions/go-framework-extension/). This framework allows us to easily deploy and operate Github Runner Webhook Router and its associated infrastructure, such as MongoDB and ingress.

Reviewer: ```suggestion
The GitHub Runner Webhook Router is a Flask application tha provides a webhook router for self-hosted GitHub Actions runners. The [GitHub Runner Webhook Router charm](https://github.com/canonical/github-runner-webhook-router/) was developed using the [12-factor Flask framework](https://canonical-charmcraft.readthedocs-hosted.com/en/stable/reference/extensions/flask-framework-extension/). This framework allows us to easily deploy and operate Github Runner Webhook Router and its associated infrastructure, such as MongoDB and ingress.
```

> The GitHub Runner Webhook Router is a Flask application tha provides a webhook router for self-hosted GitHub Actions runners. The [GitHub Runner Webhook Router charm](https://github.com/canonical/github-runner-webhook-router/) was developed using the [12-factor Flask framework](https://canonical-charmcraft.readthedocs-hosted.com/en/stable/reference/extensions/go-framework-extension/). This framework allows us to easily deploy and operate Github Runner Webhook Router and its associated infrastructure, such as MongoDB and ingress.

Reviewer 2: ```suggestion
The GitHub Runner Webhook Router is a Flask application that provides a webhook router for self-hosted GitHub Actions runners. The [GitHub Runner Webhook Router charm](https://github.com/canonical/github-runner-webhook-router/) was developed using the [12-factor Flask framework](https://canonical-charmcraft.readthedocs-hosted.com/en/stable/reference/extensions/flask-framework-extension/). This framework allows us to easily deploy and operate the Github Runner Webhook Router and its associated infrastructure, such as MongoDB and ingress.
```

It leverages the [12-factor](https://canonical-12-factor-app-support.readthedocs-hosted.com/en/latest/) support to pack a [Flask](https://flask.palletsprojects.com/) application providing the functionality for the webhook router.

For a complete view on the architecture of a 12-factor charm, refer to the [12-factor architecture documentation](https://canonical-12-factor-app-support.readthedocs-hosted.com/en/latest/explanation/charm-architecture/).

## Charm architecture diagram

Below is a diagram of the application architecture of the GitHub Runner Webhook Router charm.

```mermaid
C4Container

System_Boundary(webhookroutercharm, "Github Runner Webhook Router Charm") {

    Container_Boundary(webhookrouter_container, "Github Runner Webhook Router Workload Container") {

>         Component(webhookrouter_core, "Webhook Router", "Flask Application", "Serves web requests, publish messages.")

Reviewer 2: ```suggestion
        Component(webhookrouter_core, "Webhook Router", "Flask Application", "Serves web requests, publishes messages")
```

```bash
NAME                                READY   STATUS    RESTARTS        AGE
github-runner-webhook-router-0      2/2     Running   0               3d14h
```

This shows there are 2 containers - the one named above, as well as a container for the charm code itself.

And if you run `kubectl describe pod github-runner-webhook-router-0`, all the containers will have a command ```/charm/bin/pebble```. That's because Pebble is responsible for the processes' startup as explained above. 

## OCI images

We use [Rockcraft](https://canonical-rockcraft.readthedocs-hosted.com/en/latest/) to build the OCI image for the Github Runner Webhook Router charm.
The image is defined in [github-runner-webhook-router rock](https://github.com/canonical/github-runner-webhook-router/blob/main/rockcraft.yaml).
The rock and the charm are published to [Charmhub](https://charmhub.io/), the official repository of charms.

> See more: [How to publish your charm on Charmhub](https://canonical-charmcraft.readthedocs-hosted.com/en/stable/howto/manage-charms/#publish-a-charm-on-charmhub)

## Metrics

> Metrics are provider by the workload container at the `/metrics` endpoint at port 9102. See [flask framework](https://documentation.ubuntu.com/rockcraft/en/latest/reference/extensions/flask-framework/) for

Reviewer 2: ```suggestion
Metrics are provider by the workload container at the `/metrics` endpoint at port 9102. See the [Flask framework reference](https://documentation.ubuntu.com/rockcraft/en/latest/reference/extensions/flask-framework/) for
```
Nitpick

# Charm architecture

> The GitHub Runner Webhook Router is a Flask application tha provides a webhook router for self-hosted GitHub Actions runners.The [GitHub Runner Webhook Router charm](https://github.com/canonical/github-runner-webhook-router/) was developed using the [12-factor Go framework](https://canonical-charmcraft.readthedocs-hosted.com/en/stable/reference/extensions/go-framework-extension/). This framework allows us to easily deploy and operate Github Runner Webhook Router and its associated infrastructure, such as MongoDB and ingress.

Reviewer: ```suggestion
The GitHub Runner Webhook Router is a Flask application tha provides a webhook router for self-hosted GitHub Actions runners.The [GitHub Runner Webhook Router charm](https://github.com/canonical/github-runner-webhook-router/) was developed using the [12-factor Flask framework](https://canonical-charmcraft.readthedocs-hosted.com/en/stable/reference/extensions/flask-framework-extension/). This framework allows us to easily deploy and operate Github Runner Webhook Router and its associated infrastructure, such as MongoDB and ingress.
```

Author: ouch!
